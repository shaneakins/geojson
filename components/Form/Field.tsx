import { FieldTypes } from 'types/'

const Field = ({ label, name, register, required, validate }: FieldTypes) => (
  <label>
    <input
      type='text'
      placeholder={label}
      {...register(name, { required: ' This is required', validate })}
    />
  </label>
)

export default Field
