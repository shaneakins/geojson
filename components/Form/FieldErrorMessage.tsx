const style: React.CSSProperties = {
  marginTop: '4px',
  color: 'red',
  fontSize: '10pt',
}

type Props = { children?: React.ReactNode }

const ErrorMessage = ({ children }: Props) => {
  return <p style={style}>{children}</p>
}

export default ErrorMessage
