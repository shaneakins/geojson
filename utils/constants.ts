export const MIN_MAX_RANGE = 0.0125

export const VALID_RANGE = {
  MIN_LAT: -90 + MIN_MAX_RANGE,
  MIN_LNG: -180 + MIN_MAX_RANGE,
  MAX_LAT: 90 - MIN_MAX_RANGE,
  MAX_LNG: 180 - MIN_MAX_RANGE,
}
