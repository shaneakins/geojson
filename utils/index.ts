export const isLatValid = (lat: number) => lat >= -90 && lat <= 90
export const isLongValid = (long: number) => long >= -180 && long <= 180
