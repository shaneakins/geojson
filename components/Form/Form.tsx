import { useForm, SubmitHandler } from 'react-hook-form'
import FieldErrorMessage from './FieldErrorMessage'
import Field from './Field'
import { isLatValid, isLongValid } from '../../utils/utils'
import { VALID_RANGE } from '../../utils/constants'
import { UserSubmitFormShort } from 'types/'

type Props = {
  submitData: SubmitHandler<UserSubmitFormShort>
}

const Form = ({ submitData }: Props) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<UserSubmitFormShort>()

  return (
    <form onSubmit={handleSubmit(submitData)}>
      <div>
        <Field
          label='Latitude'
          name='lat'
          register={register}
          required
          validate={isLatValid}
        />
        {errors.lat && errors.lat.type === 'validate' && (
          <FieldErrorMessage>
            {`Latitude must be between ${VALID_RANGE.MIN_LAT} and ${VALID_RANGE.MAX_LAT}`}
          </FieldErrorMessage>
        )}
        {errors.lat && errors.lat.message && (
          <FieldErrorMessage>Please enter a latitude</FieldErrorMessage>
        )}
      </div>
      <div>
        <Field
          label='Longitude'
          name='lng'
          register={register}
          required
          validate={isLongValid}
        />
        {errors.lng && errors.lng.type === 'validate' && (
          <FieldErrorMessage>
            {`Longitude must be between ${VALID_RANGE.MIN_LNG} and ${VALID_RANGE.MAX_LNG}`}
          </FieldErrorMessage>
        )}
        {errors.lng && errors.lng.message && (
          <FieldErrorMessage>Please enter a longitude</FieldErrorMessage>
        )}
      </div>
      <input type='submit' value='Plot' />
    </form>
  )
}

export default Form
