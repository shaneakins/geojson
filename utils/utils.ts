import { VALID_RANGE } from './constants'

export const isLatValid = (lat: number) =>
  lat >= VALID_RANGE.MIN_LAT && lat <= VALID_RANGE.MAX_LAT
export const isLongValid = (long: number) =>
  long >= VALID_RANGE.MIN_LNG && long <= VALID_RANGE.MAX_LNG
export const randomId = Math.random().toString(36).slice(2)
