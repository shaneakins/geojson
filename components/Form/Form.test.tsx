import React from 'react'
import { screen, render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import Form from './Form'
import userEvent from '@testing-library/user-event'
import { VALID_RANGE } from '../../utils/constants'

const submitData = jest.fn()
const errorMsg = {
  lat: `Latitude must be between ${VALID_RANGE.MIN_LAT} and ${VALID_RANGE.MAX_LAT}`,
  lng: `Longitude must be between ${VALID_RANGE.MIN_LNG} and ${VALID_RANGE.MAX_LNG}`,
}

describe('RecipeForm', () => {
  it('should render the basic fields', () => {
    const { container } = render(<Form submitData={submitData} />)
    expect(screen.queryByPlaceholderText(/latitude/i)).toBeInTheDocument()
    expect(
      screen.queryByPlaceholderText(/longitude/i)

      // screen.getByRole('textbox', { name: /latitude/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /plot/i })).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
  it('should not show error when latitude is within allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lat = screen.queryByPlaceholderText(/latitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lat, '0')
    userEvent.click(submit)
    await waitFor(() => {
      expect(screen.queryByText(errorMsg.lat)).not.toBeInTheDocument()
    })
  })
  it('should show error when latitude is greater than allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lat = screen.queryByPlaceholderText(/latitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lat, '999')
    userEvent.click(submit)
    expect(await screen.findByText(errorMsg.lat)).toBeInTheDocument()
  })
  it('should show error when latitude is less than allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lat = screen.queryByPlaceholderText(/latitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lat, '-999')
    userEvent.click(submit)
    expect(await screen.findByText(errorMsg.lat)).toBeInTheDocument()
  })
  it('should not show error when longitude is within allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lng = screen.queryByPlaceholderText(/longitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lng, '0')
    userEvent.click(submit)
    await waitFor(() => {
      expect(screen.queryByText(errorMsg.lng)).not.toBeInTheDocument()
    })
  })
  it('should show error when longitude is greater than allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lng = screen.queryByPlaceholderText(/longitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lng, '999')
    userEvent.click(submit)
    expect(await screen.findByText(errorMsg.lng)).toBeInTheDocument()
  })
  it('should show error when longitude is less than allowed range', async () => {
    render(<Form submitData={submitData} />)
    const lng = screen.queryByPlaceholderText(/longitude/i) as Element
    const submit = screen.getByRole('button', { name: /plot/i })
    userEvent.type(lng, '-999')
    userEvent.click(submit)
    expect(await screen.findByText(errorMsg.lng)).toBeInTheDocument()
  })
})
