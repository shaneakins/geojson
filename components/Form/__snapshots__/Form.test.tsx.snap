// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeForm should render the basic fields 1`] = `
<div>
  <form>
    <div>
      <label>
        <input
          name="lat"
          placeholder="Latitude"
          type="text"
        />
      </label>
    </div>
    <div>
      <label>
        <input
          name="lng"
          placeholder="Longitude"
          type="text"
        />
      </label>
    </div>
    <input
      type="submit"
      value="Plot"
    />
  </form>
</div>
`;
